cranelift: Add LibCalls to the interpreter

afonso360 opened the pull request:
👋 Hey,

This does not fix any of the issues reported by the fuzzer so far, but I think that is just a coincidence, since at some point it will generate a libcall and we cannot handle those either.

Edit: Actually this may have already been reported, but have returned the same error as  #4758 so it wasn't reported separately.

This allows the interpreter to register libcall handlers and invoke them when called.

I switched the fuzzer generated libcall from Udiv to Ishl, reason being that we don't want libcalls that can trap. Both functions have identical signatures so we shouldn't have any input format changes to the fuzzer.

afonso360 commented on fuzz/fuzz_targets/cranelift-fuzzgen.rs:
```diff
@@ -56,7 +59,13 @@ fuzz_target!(|testcase: TestCase| {
-        let state = InterpreterState::default().with_function_store(env);
+        let state = InterpreterState::default()
+            .with_function_store(env)
+            .with_libcall(LibCall::IshlI64, &|args| match &args[..] {
+                [DataValue::I64(_), DataValue::I64(b)] if *b >= 63 => smallvec![DataValue::I64(0)],
+                [DataValue::I64(a), DataValue::I64(b)] => smallvec![DataValue::I64(a.shl(b))],
+                _ => unreachable!(),
+            });
```
I saw some sources online that point to this being the correct behaviour, but nothing definitive. Does anyone know if this is correct?

afonso360 commented on fuzz/fuzz_targets/cranelift-fuzzgen.rs:
```diff
@@ -56,7 +59,13 @@ fuzz_target!(|testcase: TestCase| {
-        let state = InterpreterState::default().with_function_store(env);
+        let state = InterpreterState::default()
+            .with_function_store(env)
+            .with_libcall(LibCall::IshlI64, &|args| match &args[..] {
+                [DataValue::I64(_), DataValue::I64(b)] if *b > 63 => smallvec![DataValue::I64(0)],
+                [DataValue::I64(a), DataValue::I64(b)] => smallvec![DataValue::I64(a.shl(b))],
+                _ => unreachable!(),
+            });
```
I saw some sources online that point to this being the correct behaviour, but nothing definitive. Does anyone know if this is correct?

afonso360 replied:
I think this is undefined behaviour, so I've added trap codes to the libcall handlers and reverted back to div

jameysharp replied:
I'm not sure I understand. We've recently defined shifts to be computed modulo the bit width. So I think `Ishl64` is a fine choice, but it should evaluate to `a << (b % 64)` for all values of `b`. That in turn means that we don't need to check the preconditions on Rust's implementation of left-shift: it would panic in debug builds if the shift amount is 64 or higher, but we'll never hit that case.

afonso360 replied:
These libcalls end up being linked to the platform's runtime library if the interpreted version succeeds, so we need to make sure that they match the behaviour agreed by the runtime libraries and not our own.

We also can't really panic here since it would cause a fuzz error.

afonso360 replied:
Another option would be to define our own symbol for the JIT as well, and then we control both sides, the interpreted version and the compiled version.

afonso360 commented on cranelift/interpreter/src/interpreter.rs:
```diff
@@ -191,9 +192,12 @@ pub enum HeapInit {
     FromBacking(HeapBacking),
 }
 
+pub type LibCallHandler<'a, V> = &'a dyn Fn(SmallVec<[V; 1]>) -> SmallVec<[V; 1]>;
```
I think it might also be a good idea to return `Result<SmallVec<[V; 1]>, TrapCode>` and allow these to trap controllably.

jameysharp commented:
Huh. I was going to suggest that if there are libcalls with undefined behavior on some inputs then we should be limiting what we pass them to ensure that we only use them in ways consistent with the semantics we want. But I went to check how libcalls like `UdivI64` or `IshlI64` are used, and as far as I can tell, they aren't. We should probably remove them, yeah?

These libcalls are constructed in Rust source:
- `ElfTlsGetAddr`
- `ElfTlsGetOffset`
- `Memcmp`
- `Memcpy`
- `Memmove`
- `Memset`
- `Probestack`

While these are constructed in the x64 ISLE rules (but not any other backend?):
- `CeilF32`
- `CeilF64`
- `FloorF32`
- `FloorF64`
- `FmaF32`
- `FmaF64`
- `NearestF32`
- `NearestF64`
- `TruncF32`
- `TruncF64`

`LibCall::for_inst` would construct some other libcalls, but apparently its only use was removed last year in bae4ec642798ff448ca88eab771b6fcea71e7884, so I've excluded it from the above lists.

afonso360 commented:
> But I went to check how libcalls like UdivI64 or IshlI64 are used, and as far as I can tell, they aren't. We should probably remove them, yeah?

That sounds like a good idea.

> While these are constructed in the x64 ISLE rules (but not any other backend?):

Yeah, x86 doesn't have native instructions for these without extensions, so we have to lower them as libcalls if some extensions are disabled. I only know about AArch64 but they have equivalents for these in the base ISA. Not sure about s390x but i assume its the same

We can emit something like `CeilF32` or any other really. The only downside being that it puts us in the "Changes fuzzer input do not merge" category. I'm okay with that.

I think the float libcalls don't trap at all.

afonso360 commented:
I've changed this in a few ways, and since we are going to change the input of the fuzzer we might as well do it properly and generate multiple libcalls, so we now do that.

I've also removed the unused libcalls and a few accessory bits.

jameysharp commented on fuzz/fuzz_targets/cranelift-fuzzgen.rs:
```diff
@@ -51,16 +53,30 @@ fn run_in_host(compiled_fn: &CompiledFunction, args: &[DataValue]) -> RunResult
-    };
+    let state = InterpreterState::default()
+        .with_function_store(env)
+        .with_libcall_handler(&|libcall: LibCall, args: SmallVec<[DataValue; 1]>| {
+            use LibCall::*;
+            Ok(smallvec![match (libcall, &args[..]) {
+                (CeilF32, [DataValue::F32(a)]) => DataValue::F32(a.ceil()),
+                (CeilF64, [DataValue::F64(a)]) => DataValue::F64(a.ceil()),
+                (FloorF32, [DataValue::F32(a)]) => DataValue::F32(a.floor()),
+                (FloorF64, [DataValue::F64(a)]) => DataValue::F64(a.floor()),
+                (TruncF32, [DataValue::F32(a)]) => DataValue::F32(a.trunc()),
+                (TruncF64, [DataValue::F64(a)]) => DataValue::F64(a.trunc()),
+                _ => unreachable!(),
+            }])
+        });
```
What do you think of defining this function in `function_generator.rs` next to the array of allowed libcalls?

afonso360 replied:
The only reason I'm hesitant to do this is that `fuzzgen` is supposed to be executor agnostic (i.e. icache needs none of this), so it feels kinda weird to have a interpreter specific function there.

jameysharp replied:
Ah, that makes sense. But this is weird for the same reason that the function references are: only outer fuzz target knows which functions and which libcalls are okay to call. So I think the set of allowed libcalls (and functions) should be provided by the fuzz target. But I also don't think that needs to block merging this PR, it's just something I think we should revisit soon.

afonso360 replied:
>So I think the set of allowed libcalls (and functions) should be provided by the fuzz target. 

I like that! 

> But I also don't think that needs to block merging this PR, it's just something I think we should revisit soon.

👍

jameysharp commented on cranelift/interpreter/src/interpreter.rs:
```diff
@@ -191,9 +192,13 @@ pub enum HeapInit {
     FromBacking(HeapBacking),
 }
 
+pub type LibCallHandler<'a, V> =
+    &'a dyn Fn(LibCall, SmallVec<[V; 1]>) -> Result<SmallVec<[V; 1]>, TrapCode>;
```
Since this `SmallVec` type ends up appearing in other modules, I'd suggest introducing a type alias for it. That way it's easy later either to change it to a regular `Vec`, or to put more elements in the inline array.

Also, `&dyn Fn` is probably more general than necessary. A closure that doesn't capture anything from its environment is compatible with a `fn` type, and I'd be surprised if interpreter libcall handlers ever needed context.

```suggestion
pub type LibCallValues<V> = SmallVec<[V; 1]>;
pub type LibCallHandler<V> = fn(LibCall, LibCallValues<V>) -> Result<LibCallValues<V>, TrapCode>;
```

jameysharp reviewed (approved):
This looks great! I was briefly sad that `Opcode::Call` handling got so much more complicated, but it all looks like useful complexity to me, so :+1:

When we're ready to merge fuzzgen changes, I'm in favor of this one. I do have a couple suggestions, but I'd also merge it as-is.
